### DISCUSSION TITLE
How to make EntityManager's flush method to flush only certain entity instances?

### DISCUSSION THREAD

**Discussion by MiguelAlcaino** [symfony/symfony#52482] [Q&A]
Hi!
My issue is that EntityManager's flush method is writing in the database all the operations buffered by the persist method. No problem with that, BUT sometimes changes get written in other context (deeper layers of my symfony app) without me knowing about it. Let me show you an example:

```php
class ServiceA{
   public function __construct(private UserRepository $userRepository, EntityManagerInterface $em){}
   
   public function foo(){
      $user = $this->userRepository->find(1);
      $user->setName('newName');
      $this->em->persist($user);
      $this->em->flush();
   }
}

class ServiceB{
    public function __construct(
        private VisitRepository $visitRepository,
        private ServiceA $serviceA,
        private EntityManagerInteface $em
    ){}

    public function bar(){
        $visit = $this->visitRepository->find(2);
        $visit->setSomething(2);
        $this->visitRepository->persist($visit);
        $this->serviceA->foo(); // Inside here the previous persisted instances in this context got flushed by the EntityManager's flush method.

        $this->em->flush(); // In here there's nothing to flush because $visit got flushed inside ServiceA's foo method.
    }
}
```

The EntityManager flushes everything, then my problem is that sometimes deep services are flushing changes done in other layers of my app without me noticing. So far it has not been a problem for me, but I can see a design problem here (on my side).

How do you tackle this issue? Do you have any design idea to apply to sort this out. Or, is there any way to wrap only certain operations within the flush method so it only writes in the DB the operations of the context?

Ideally I would like to have this behaviour in the previous code:
```php
class ServiceB{
...
public function bar(){
        $visit = $this->visitRepository->find(2);
        $visit->setSomething(2);
        $this->visitRepository->persist($visit);
        $this->serviceA->foo(); // if this flushes something inside it should only flush entities in ServiceA's context

        $this->em->flush(); // it only flushes $visit
    }
...
}
```

I imagine something like this as a solution to accurate the operations that the Entity flushes:
```
$otherContextEntityManager->persist($bar);
$entityManagerX->flush(); // Nothing is flushed.
$thisContextEntityManager->persist($something)
$thisContextEntityManager->persist($foo);

$thisContextEntityManager->flush(); // Only $something and $foo get written in the DB

$otherContextEntityManager->persist($james);
$otherContextEntityManager->flush(); // Only $bar and $james get written in the DB.

```

Maybe to make some EntityManager as a not shared service?

**Comment by stof** [comment]
The only way would be to manage 2 different entity managers. But this, this requires making sure you never mix entities managed by each of them in the object graph (as the other entity manager would treat them as *new* objects that need to be inserted).

Generally, it is better to respect the Doctrine ORM architecture if you want to use that ORM.

My own rule is that most services are not allowed to call `flush`, reserving that to entrypoint services only (commands, controllers and messenger handlers)

**Comment by MiguelAlcaino** [comment]
Hey @stof that's exactly what I've been doin, to only call flush (and persists) on entrypoint services so it's really visible and explicit when I write in the database. Thanks for the advice, I'm glad that I'm not the only one doing that.

**Comment by yceruto** [comment]
Another approach is wrapping your domain change operations "Commands" into an ACID transaction using `EntityManagerInterface::wrapInTransaction()`.

**Reply by stof** [reply]
@yceruto this won't help for that case. `EntityManagerInterface::wrapInTransaction()` still flushes the whole unit of work. This methods solves a different use case (adding _additional_ queries that must be in the same transaction than the flush)

**Reply by MiguelAlcaino** [reply]
Hi @yceruto . I thought, some time ago, that what you are proposing would be the solution to my issue. But when I tested it, the internal `flush` invocation within the `wrapInTransaction` method still flushed previous persisted entities. :(

**Reply by yceruto** [reply]
It works for me because I never manually call the `flush` method in any service.

My personal rule involves dispatching a command message through a bus that wraps all changes within an ACID transaction. If you need to divide those changes into separate database operations, you can do so by using two different commands or by publishing domain events asynchronously to eventually achieve consistency with the previous changes in your related data.

**Reply by MiguelAlcaino** [reply]
Hey @yceruto do you have a code example regarding your approach or any link to read about placing transactions in a command message?
And when you say command message, do you mean using something like `Symfony\Component\Messenger\MessageBusInterface` ?
I'm interested in your approach.
Thanks in advance!

**Reply by yceruto** [reply]
It's already supported by Symfony Messenger. See https://github.com/symfony/doctrine-bridge/blob/7.0/Messenger/DoctrineTransactionMiddleware.php, which can be configured as one of the middleware for your command bus.

More here https://symfony.com/doc/current/messenger.html#transactional-messages-handle-new-messages-after-handling-is-done

About the approach I'm using for my own projects, you can take a look here https://github.com/yceruto/ddd-cqs-rest-api-demo, still in progress to integrate Symfony Messenger for async dispatching, but the main approach is there.

**Reply by ghost** [reply]
@yceruto  Transactions have nothing to do with flushing. A transaction only prevents database changes (outside the transaction scope, not inside) until the transaction is committed. Flushing is a Doctrine ORM concept that in the end executes SQL queries against a database to make PHP entities changes persistent.

@MiguelAlcaino 
> How to make EntityManager's flush method to flush only certain entity instances?

Every call to `flush()`, no matter where in the code, without an argument will flush the current state of UOW. Period.
You have to take it into account while making implementation decisions.

In doctrine.orm <=2 you can pass an entity to `flush($entity)`. However in v3 you won't be allowed to because it was removed. You can search their issues for the issue where it was removed - maybe there will be a justification and a discussion in the comments.

**Reply by stof** [reply]
The justification is that partial flushes are broken by design: that entity might reference others, which would just explode during the saving if they are not flushed as well. And it was the root cause of lots of weird bugs (when the changes in the unit of work get saved only partially, it forces to handle weird cases almost everywhere in the ORM)

**Comment by ghost** [comment]
You asked the same question before in doctrine/orm discussions (btw. this is where this question belongs).

**Reply by MiguelAlcaino** [reply]
Yes, I removed it from there to place it here.
